### PR TITLE
No need for reconcile if m4dapplication is in Deny state

### PR DESCRIPTION
**Pull request by shlomitk1**
May fix https://github.com/mesh-for-data/mesh-for-data/issues/708

This PR has the following changes:
- conditions are Ready, Deny, Error instead of Error, Failure (Deny is set either if access is denied, or the asset does not exist)
- error analysis to determine the appropriate condition including grpc errors
- reconcile if app is not in Ready or Deny state instead of reconcile if app is not Ready
- added a check in DenyOnRead test whether a new reconcile will be made

### CODE REVIEW

File: manager/controllers/app/m4dapplication_controller.go [hunk 1]
```diff
@@ -471,15 +471,29 @@ func (r *M4DApplicationReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+		return
+	}
+	// Unwrap the error - it wraps the original error by the connector
+	errorToCheck := err
+	if cause := errors.Cause(err); cause != nil {
+		errorToCheck = cause
+	}
+	if errStatus, fromGrpc := status.FromError(errorToCheck); fromGrpc {
+		if errStatus.Code() == codes.InvalidArgument {
+			setDenyCondition(application, assetID, errStatus.Message())
```
Comment (roee88) [inline manager/controllers/app/m4dapplication_controller.go:485]: So there is an undocumented API here that if the asset is unknown the connector must return InvalidArgument? Do we need to document it?

File: manager/controllers/app/m4dapplication_conditions.go [hunk 2]
```diff
@@ -11,31 +11,41 @@ import (
...
+	application.Status.Conditions[app.ReadyConditionIndex] = app.Condition{Type: app.ReadyCondition, Status: corev1.ConditionFalse}
 }
 
-func setCondition(application *app.M4DApplication, assetID string, msg string, fatalError bool) {
+func setErrorCondition(application *app.M4DApplication, assetID string, msg string) {
 	if len(application.Status.Conditions) == 0 {
 		resetConditions(application)
 	}
 	errMsg := "An error was received"
 	if assetID != "" {
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply manager/controllers/app/m4dapplication_conditions.go:26]: A general k8s error, an error from the plotter/blueprint (it does not map assets to helm releases)

File: manager/controllers/app/m4dapplication_conditions.go [hunk 3]
```diff
@@ -11,31 +11,41 @@ import (
...
-		ind = app.ErrorConditionIndex
-	}
-	application.Status.Conditions[ind].Status = corev1.ConditionTrue
-	application.Status.Conditions[ind].Message += errMsg
+	application.Status.Conditions[app.ErrorConditionIndex].Status = corev1.ConditionTrue
+	application.Status.Conditions[app.ErrorConditionIndex].Message += errMsg
+}
+
+func setDenyCondition(application *app.M4DApplication, assetID string, msg string) {
+	if len(application.Status.Conditions) == 0 {
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply manager/controllers/app/m4dapplication_conditions.go:36]: ok

File: manager/controllers/app/m4dapplication_conditions.go [hunk 4]
```diff
@@ -44,7 +54,17 @@ func hasError(application *app.M4DApplication) bool {
 		return false
 	}
 	return (application.Status.Conditions[app.ErrorConditionIndex].Status == corev1.ConditionTrue ||
-		application.Status.Conditions[app.FailureConditionIndex].Status == corev1.ConditionTrue)
+		application.Status.Conditions[app.DenyConditionIndex].Status == corev1.ConditionTrue)
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply manager/controllers/app/m4dapplication_conditions.go:57]: ok

File: manager/controllers/app/m4dapplication_conditions.go [hunk 5]
```diff
@@ -11,31 +11,41 @@ import (
...
 	errMsg := "An error was received"
 	if assetID != "" {
-		errMsg += " for asset " + assetID + " . "
-	}
-	if !fatalError {
-		errMsg += "If the error persists, please contact an operator.\n"
+		errMsg += " for asset " + assetID
 	}
+	errMsg += " . If the error persists, please contact an operator.\n"
 	errMsg += "Error description: " + msg + "\n"
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply manager/controllers/app/m4dapplication_conditions.go:30]: ok

File: manager/controllers/app/m4dapplication_controller.go (hunk 1 above)
Comment (shlomitk1) [reply manager/controllers/app/m4dapplication_controller.go:485]: I think we do.

Comment (roee88) [reply manager/controllers/app/m4dapplication_controller.go:485]: There is also a leakage of the abstractions here by looking at grpc specifics. That will be an issue when we move to openapi for example. Maybe the right approach is to change the DataCatalog interface to convert to `app.*` errors.  If this works for our current connector then it's fine for this PR but we will need to resolve it later.

**Review by roee88 — APPROVED** [review]

**Pull request merged**